pkg/auth/artifact_verifier.go

```diff
@@ -0,0 +1,162 @@
+//
+// If the artifact is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz
+//
+// Then its build manifest is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz.manifest
+//
+// artifact_sha: abc23456
+//
+// And its signature file is located here:
+// https://foo.bar.baz/artifacts/myapp_abc123.tar.gz.manifest.sig
+type BuildManifestVerifier struct {
+	keyring openpgp.KeyRing
+	fetcher uri.Fetcher
+	logger  *logging.Logger
+}
+
+func NewBuildArtifactVerifier(keyringPath string, fetcher uri.Fetcher, logger *logging.Logger) (*BuildManifestVerifier, error) {
+	keyring, err := LoadKeyring(keyringPath)
+	if err != nil {
+		return nil, util.Errorf("Could not load artifact verification keyring from %v: %v", keyringPath, err)
+	}
+	return &BuildManifestVerifier{
+		keyring: keyring,
+		fetcher: fetcher,
+		logger:  logger,
+	}, nil
+}
+
+// Returns an error if the stanza's artifact is not signed appropriately. Note that this
+// implementation does not use the pod manifest digest location options.
+func (b *BuildManifestVerifier) VerifyHoistArtifact(localCopy *os.File, artifactLocation string) error {
+	u, err := url.Parse(artifactLocation)
+	if err != nil {
+		return err
+	}
+	switch u.Scheme {
+	default:
+		return fmt.Errorf("%v does not have a recognized scheme, cannot verify manifest or signature", artifactLocation)
+	case "http", "https", "file":
+		dir, err := ioutil.TempDir("", "artifact_verification")
+		if err != nil {
+			return fmt.Errorf("Could not create temporary directory for manifest file: %v", err)
+		}
+		defer os.RemoveAll(dir)
+
+		manifestSrc := fmt.Sprintf("%v.manifest", artifactLocation)
+		manifestDst := filepath.Join(dir, "manifest")
+		err = b.fetcher.CopyLocal(manifestSrc, manifestDst)
+		if err != nil {
+			return fmt.Errorf("Could not download artifact manifest for %v: %v", artifactLocation, err)
+		}
+
+		signatureSrc := fmt.Sprintf("%v.sig", manifestSrc)
+		signatureDst := filepath.Join(dir, "signature")
+		if err = b.fetcher.CopyLocal(signatureSrc, signatureDst); err != nil {
+			return fmt.Errorf("Could not download manifest signature for %v: %v", artifactLocation, err)
+		}
+
+		manifestBytes, err := ioutil.ReadFile(manifestDst)
+		if err != nil {
+			return err
+		}
+		signatureBytes, err := ioutil.ReadFile(signatureDst)
+		if err != nil {
+			return err
+		}
+
+		if err = b.verifySigned(manifestBytes, signatureBytes); err != nil {
+			return err
+		}
+
+		return b.checkMatchingDigest(localCopy, manifestBytes)
+	}
+}
+
+func (b *BuildManifestVerifier) verifySigned(manifestBytes, signatureBytes []byte) error {
+	// permit an armored detached signature
+	block, err := armor.Decode(bytes.NewBuffer(signatureBytes))
+	if err == nil {
+		signatureBytes, err = ioutil.ReadAll(block.Body)
+		if err != nil {
+			return err
+		}
+	}
+	// check that the manifest was adequately signed by our signer
+	_, err = checkDetachedSignature(b.keyring, manifestBytes, signatureBytes)
```

Comment:
do you need to check who the signer was here? or just that it was signed by something in the keyring